tests/Aspire.Hosting.Tests/Dashboard/ResourcePublisherTests.cs

```diff
@@ -131,6 +131,35 @@ public async Task DeletesRemoveFromSnapshot()
         await cts.CancelAsync();
     }
 
+    [Fact]
+    public async Task CancelledSubscriptionIsCleanedUp()
+    {
+        ResourcePublisher publisher = new(CancellationToken.None);
+        CancellationTokenSource cts = new();
+        var called = false;
+
+        var (_, subscription) = publisher.Subscribe();
+
+        var task = Task.Run(async () =>
+        {
+            await foreach (var item in subscription.WithCancellation(cts.Token).ConfigureAwait(false))
+            {
+                // We should only loop one time.
+                Assert.False(called);
+                called = true;
+
+                // Now we've received something, cancel.
+                await cts.CancelAsync();
+            }
+        });
+
+        // Push through an update.
+        await publisher.IntegrateAsync(CreateResourceSnapshot("A"), ResourceSnapshotChangeType.Upsert).ConfigureAwait(false);
+
+        // Let the subscriber exit.
+        await task;
```

Comment:
I believe this will be flaky. GetBatch exits if the token is canceled in some situations but can also throw OperationCanceledException from WaitToReadAsync inside GetBatch. That then surfaces here. Either you need to put a try/catch around the await in the task, or a try/catch inside GetBatch.

A decision needs to be made about whether exiting out of async enumerables with cancellation consistently throws OCE or not.

Reply:
> A decision needs to be made about whether exiting out of async enumerables with cancellation consistently throws OCE or not.

I don't have a strong opinion either:

* Throwing OCE is more technically correct for cancel behavior
* But on the other hand, not throwing OCE and having to catch it everywhere is nice

But I think being inconsistent like we are now isn't good.

Reply:
> I believe this will be flaky. The task could sometime throw OperationCanceledException.

The cancellation happens within the task's loop in a deterministic way, so I don't expect this to be flakey. Awaiting `CancelAsync` should ensure that everything is set for the next async enumerator iteration to loop around.

Using VS Test Explorer's "Run until failure" option for showed no failure after many iterations:

![image](https://github.com/dotnet/aspire/assets/350947/ef1f0374-e66c-4e4e-86b0-371a441b0e83)

> A decision needs to be made about whether exiting out of async enumerables with cancellation consistently throws OCE or not.

I'm not sure we can control that without wrapping every `await foreach` sequence, as if cancelled during the body, the `foreach` loop breaks cleanly. Can we influence the generated code? Or would you want to wrap this somehow.

Either way, it can be done separately across the repo, so please file an issue if you would like to see it addressed.

Reply:
> The cancellation happens within the task's loop in a deterministic way

Ah yes. If the cancel happened outside of the task then it would be flaky.

> I'm not sure we can control that without wrapping every await foreach sequence, as if cancelled during the body, the foreach loop breaks cleanly. Can we influence the generated code?

An error is thrown from await foreach if an error is thrown from the method that returns `IAsyncEnumerable`. All that is required is adding a catch inside the method.

e.g.

```cs
                using var linked = CancellationTokenSource.CreateLinkedTokenSource(enumeratorCancellationToken, cancellationToken);

                try
                {
                    await foreach (var batch in channel.GetBatches(linked.Token).ConfigureAwait(false))
                    {
                        yield return batch;
                    }
                }
                catch (OperationCancellationException ex) when (ex.CancellationToken == linked.Token)
                {
                    // nom nom nom
                }
                finally
                {
                    ImmutableInterlocked.Update(ref _outgoingChannels, static (set, channel) => set.Remove(channel), channel);
                }
```

Reply:
https://github.com/dotnet/aspire/issues/1618